### ISSUE TITLE
`uninlined_format_args`  should have stayed in `pedantic`

### ISSUE THREAD

**Issue by IceSentry** [rust-lang/rust-clippy#15151]
### Description

Format args currently don't support field access. This means you are forced to use uninlined format args when you need field access. Since I commonly mix field access and variables it means I would be forced to mix and match inlined and uninlined format args all over the place. For this reason I don't inline anything because I would always end up refactoring my format args.

Moving this lint to `style` now means we are forced by default to mix 2 formatting styles and constantly refactor format strings when needing field access. This feels like a very pedantic thing to warn about. I don't understand why it was moved to `style`.

Until field access is possible inside format args warning on this in every codebase is going to generate a lot of noise for very little gain.

Inlining everything can also generate really long strings which breaks `rustfmt` which is not great. There's also issues like this where inlining everything  isn't really nice https://github.com/rust-lang/rust-clippy/issues/14686

I think this lint was moved to `style` prematurely.

### Version

```text
rust 1.88
```

**Comment by pitaj** [comment]
By default, the lint does not trigger on anything but when only variables are passed directly. This has changed since the initial enablement of the lint, so I understand your concern - but it's no longer an issue.

[playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=88b063ae0bce0199ae94f3680e060fa3)

So I'm going to close this for now. If you have a specific snippet that bothers you, please provide it and reopen the issue.

**Comment by wks** [comment]
I agree.  I think `uninlined_format_args` should have stayed in `pedantic` if not removed completely.

The "Why is this bad" part given by https://rust-lang.github.io/rust-clippy/master/index.html#uninlined_format_args is:

> Non-inlined code is slightly more difficult to read and understand, as it requires arguments to be matched against the format string. The inlined syntax, where allowed, is simpler.

The argument "Non-inlined code is slightly more difficult to read and understand" is not always invalid.  The readability of a format string is determined by many aspects.  One important part is *whether human eyes can capture the overall structure at a single glance*.  For example,

```diff
-log::info!("Job competed in {time} seconds");
+log::info!("Job completed in {} seconds", time);
```

Both are OK because they are sufficiently short.

```diff
-log::info!("New record added.  foo: {} (in {} seconds), bar: {}, baz: {}", foo, foo_time, bar, baz);
+log::info!("New record added.  foo: {foo} (in {foo_time} seconds), bar: {bar}, baz: {baz}");
```

This may be the intended case that `uninlined_format_args` is trying to capture.  This format string has a somewhat complex structure, and human is having a hard time matching the args with the placeholders.

```diff
-log::info!("Finished.  foo time: {}, bar time: {}, total time: {}",
-        process_foo_time_milliseconds,
-        process_bar_time_milliseconds,
-        process_foo_time_milliseconds + process_bar_time_milliseconds);
+log::info!("Finished.  foo time: {process_foo_time_milliseconds}, bar time: {process_bar_time_milliseconds}, total time: {}",
+        process_foo_time_milliseconds + process_bar_time_milliseconds);
```

(Note that the variable names are long, but they are meaningful and unambiguous in their context.  There may be similarly named variables such as `process_foo_time: f64`.  We shouldn't change their names just for formatting.)

This is where things start to become problematic.  The original format string `"Finished.  foo time: {}, bar time: {}, total time: {}"` is very clear in terms of structure.  Humans know it has three placeholders and what they are for (foo time, bar time and total time, respectively).

In contrast, the inlined format `"Finished.  foo time: {process_foo_time_milliseconds}, bar time: {process_bar_time_milliseconds}, total time: {}"` is so long that it doesn't even fit into the line width of the text editor.  Humans will have to look back and forth and skip those variable names just to grasp the overall structure.  And it is currently impossible to inline complex expressions like `a + b` and `a.foo()` which is often needed in log outputs.  We can't hoist costly expressions like `a.foo()` before logging statements like `log::debug!` and assign it into a variable because that will force the expression to be evaluated when the log level is disabled.

And the un-inlined format string is more similar to the output structure than the inlined format.  The actual times may be just a few digits (like `foo time: 42`) which is similar to `foo time: {}` in length, not `foo time: {process_foo_time_milliseconds}`.

Actually, the formatting macro provides us one more option that balances terseness and match-ability.

```rust
log::info!("Finished.  foo time: {f}, bar time: {b}, total time: {t}",
        f = process_foo_time_milliseconds,
        b = process_bar_time_milliseconds,
        t = process_foo_time_milliseconds + process_bar_time_milliseconds);
```

But the `uninlined_format_args` lint forbids this format, too, as shown on https://rust-lang.github.io/rust-clippy/master/index.html#uninlined_format_args

I think the lint `uninlined_format_args` is *pointless*.  For large projects with many logging statements, it seldom improves the readability.  It is also denying useful mechanisms Rust has [provided](https://doc.rust-lang.org/std/fmt/index.html) us to satisfy many different use cases developers would encounter in real-world projects.  The goal of this lint is obviously in conflict with that of the [std::fmt](https://doc.rust-lang.org/std/fmt/index.html) module which provides powerful mechanisms.

**Comment by your-diary** [comment]
IMHO, the lint should be in `pedantic` *forever*.

- (As pointed out by @wks) the problem isn't limited to field accesses but is generally applicable to any expressions (e.g. `a + b`).

- It affects the readability too much. I'm even feeling like I'm being forced a personal preference. I respect those who prefer inlined style, but *I* don't like the style so I want to be given an option to choose which to use via *opt-in*.

**Comment by dgagn** [comment]
Also, refactors such as renaming makes this unbearable since we need to find every references to the variable name and modify it manually. It should not be the default for sure!
